File: base/src/main/java/com/zhangteng/base/base/BasePopupWindow.java
package com.zhangteng.base.base;

import android.app.Activity;
import android.content.Context;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.support.constraint.ConstraintLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.PopupWindow;

import com.zhangteng.base.R;

/**
 * Created by swing on 2018/9/6.
 */
public abstract class BasePopupWindow extends PopupWindow {
    protected LinearLayout clTitle;
    protected ConstraintLayout clContent;
    protected LinearLayout clButton;
    protected View parent;

    protected OnCancelClickListener onCancelClickListener;
    protected OnConfirmClickListener onConfirmClickListener;

    public BasePopupWindow(Context context) {
        super(context);
        initView(context);
    }

    private void initView(Context context) {
        parent = LayoutInflater.from(context).inflate(R.layout.self_base_popupwindow, null);
        clTitle = parent.findViewById(R.id.self_base_popupwindow_title);
        clContent = parent.findViewById(R.id.self_base_popupwindow_content);
        clButton = parent.findViewById(R.id.self_base_popupwindow_button);

        if (getSelfTitleView() != 0) {
            LayoutInflater.from(context).inflate(getSelfTitleView(), clTitle, true);
        }

        if (getSelfContentView() != 0) {
            LayoutInflater.from(context).inflate(getSelfContentView(), clContent, true);
        }

        if (getSelfButtonView() != 0) {
            LayoutInflater.from(context).inflate(getSelfButtonView(), clButton, true);
        }
        initView(parent);
        this.setContentView(parent);

        //设置高
        this.setHeight(WindowManager.LayoutParams.WRAP_CONTENT);
        //设置宽
        this.setWidth(WindowManager.LayoutParams.MATCH_PARENT);
        //设置SelectPicPopupWindow弹出窗体可点击
        this.setFocusable(true);
        //设置PopupWindow可触摸
        this.setTouchable(true);
        //设置非PopupWindow区域是否可触摸
        this.setOutsideTouchable(true);
        //实例化一个ColorDrawable颜色为半透明
        ColorDrawable dw = new ColorDrawable(0x00000000);
        //设置SelectPicPopupWindow弹出窗体的背景
        this.setBackgroundDrawable(dw);
        //设置SelectPicPopupWindow弹出窗体动画效果
        this.setAnimationStyle(R.style.showAsDropDown);
        //防止被虚拟导航栏阻挡
        this.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
    }

    public abstract int getSelfTitleView();

    public abstract int getSelfContentView();

    public abstract int getSelfButtonView();

    public abstract void initView(View view);

    /**
     * 对第一个按钮设置确定监听器
     */
    public void setOnCancelClickListener(OnCancelClickListener onCancelClickListener) {
        this.onCancelClickListener = onCancelClickListener;
        if (clButton.getChildCount() > 0 && clButton.getChildAt(0) instanceof ViewGroup) {
            if (((ViewGroup) clButton.getChildAt(0)).getChildAt(0) != null) {
                ((ViewGroup) clButton.getChildAt(0)).getChildAt(0).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (BasePopupWindow.this.onCancelClickListener != null) {
                            BasePopupWindow.this.onCancelClickListener.onCancel(v);
                        }
                    }
                });
            }
        }
    }

    /**
     * 对最后一个按钮设置确定监听器
     */
    public void setOnConfirmClickListener(OnConfirmClickListener onConfirmClickListener) {
        this.onConfirmClickListener = onConfirmClickListener;
        if (clButton.getChildCount() > 0 && clButton.getChildAt(0) instanceof ViewGroup) {
            if (((ViewGroup) clButton.getChildAt(0)).getChildAt(((ViewGroup) clButton.getChildAt(0)).getChildCount() - 1) != null) {
                ((ViewGroup) clButton.getChildAt(0)).getChildAt(((ViewGroup) clButton.getChildAt(0)).getChildCount() - 1).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (BasePopupWindow.this.onConfirmClickListener != null) {
                            BasePopupWindow.this.onConfirmClickListener.onConfirm(v);
                        }
                    }
                });
            }
        }
    }

    @Override
    public View getContentView() {
        return super.getContentView();
    }

    @Override
    public void setContentView(View view) {
        super.setContentView(view);
    }

    @Override
    public void setAnimationStyle(int animationStyle) {
        super.setAnimationStyle(animationStyle);
    }

    @Override
    public void setBackgroundDrawable(Drawable background) {
        super.setBackgroundDrawable(background);
    }

    @Override
    public void showAsDropDown(View view) {
        super.showAsDropDown(view);
    }

    public void setDropDown() {
        //设置SelectPicPopupWindow弹出窗体动画效果
        this.setAnimationStyle(R.style.showAsDropDown);
    }

    public void setDropUp() {
        this.setAnimationStyle(R.style.showAsDropUp);
    }

    @Override
    public void showAtLocation(View parent, int gravity, int x, int y) {
        super.showAtLocation(parent, gravity, x, y);
    }

    @Override
    public void showAsDropDown(View anchor, int xoff, int yoff) {
        super.showAsDropDown(anchor, xoff, yoff);
    }

    @Override
    public void showAsDropDown(View anchor, int xoff, int yoff, int gravity) {
        super.showAsDropDown(anchor, xoff, yoff, gravity);
    }

    /**
     * 窗口变暗
     */
    public void showBlackWindowBackground(Activity activity) {
        WindowManager.LayoutParams lp = activity.getWindow()
                .getAttributes();
        lp.alpha = 0.4f;
        activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);
        activity.getWindow().setAttributes(lp);
    }

    public void dismissBlackWindowBackground(Activity activity) {
        WindowManager.LayoutParams lp = activity.getWindow()
                .getAttributes();
        lp.alpha = 1f;
        activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);
        activity.getWindow().setAttributes(lp);
    }

    /**
     * 取消监听器
     */
    public interface OnCancelClickListener {
        void onCancel(View view);
    }

    /**
     * 确定监听器
     */
    public interface OnConfirmClickListener {
        void onConfirm(View view);
    }
}

File: base/src/main/java/com/zhangteng/base/base/BasePopupWindow.kt
package com.zhangteng.base.base

import android.annotation.SuppressLint
import android.app.Activity
import android.content.*
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.view.*
import android.widget.LinearLayout
import android.widget.PopupWindow
import com.zhangteng.base.R

/**
 * Created by swing on 2018/9/6.
 */
abstract class BasePopupWindow(context: Context?) : PopupWindow(context) {
    protected lateinit var parent: View
    protected var mTitle: LinearLayout? = null
    protected var mContent: LinearLayout? = null
    protected var mButton: LinearLayout? = null
    var onCancelClickListener: OnCancelClickListener? = null
        set(value) {
            field = value
            if (mButton != null && mButton!!.childCount > 0 && mButton!!.getChildAt(0) is ViewGroup) {
                if ((mButton?.getChildAt(0) as ViewGroup).getChildAt(0) != null) {
                    (mButton?.getChildAt(0) as ViewGroup).getChildAt(0).setOnClickListener { v ->
                        if (this@BasePopupWindow.onCancelClickListener != null) {
                            this@BasePopupWindow.onCancelClickListener!!.onCancel(v)
                        }
                    }
                }
            }
        }
    var onConfirmClickListener: OnConfirmClickListener? = null
        set(value) {
            field = value
            if (mButton != null && mButton!!.childCount > 0 && mButton!!.getChildAt(0) is ViewGroup) {
                if ((mButton!!.getChildAt(0) as ViewGroup).getChildAt((mButton!!.getChildAt(0) as ViewGroup).childCount - 1) != null) {
                    (mButton!!.getChildAt(0) as ViewGroup).getChildAt((mButton!!.getChildAt(0) as ViewGroup).childCount - 1)
                        .setOnClickListener { v ->
                            if (this@BasePopupWindow.onConfirmClickListener != null) {
                                this@BasePopupWindow.onConfirmClickListener!!.onConfirm(v)
                            }
                        }
                }
            }
        }

    @SuppressLint("InflateParams")
    protected fun initView(context: Context?) {
        parent = LayoutInflater.from(context).inflate(R.layout.self_base_popup, null)
        mTitle = parent.findViewById(R.id.self_base_popup_title)
        mContent = parent.findViewById(R.id.self_base_popup_content)
        mButton = parent.findViewById(R.id.self_base_popup_button)
        if (getSelfTitleView() != 0) {
            LayoutInflater.from(context).inflate(getSelfTitleView(), mTitle, true)
        }
        if (getSelfContentView() != 0) {
            LayoutInflater.from(context).inflate(getSelfContentView(), mContent, true)
        }
        if (getSelfButtonView() != 0) {
            LayoutInflater.from(context).inflate(getSelfButtonView(), mButton, true)
        }
        this.contentView = parent

        //设置高
        this.height = WindowManager.LayoutParams.WRAP_CONTENT
        //设置宽
        this.width = WindowManager.LayoutParams.MATCH_PARENT
        //设置SelectPicPopupWindow弹出窗体可点击
        this.isFocusable = true
        //设置PopupWindow可触摸
        this.isTouchable = true
        //设置非PopupWindow区域是否可触摸
        this.isOutsideTouchable = true
        //实例化一个ColorDrawable颜色为半透明
        val dw = ColorDrawable(0x00000000)
        //设置SelectPicPopupWindow弹出窗体的背景
        setBackgroundDrawable(dw)
        //默认向下弹出
        this.animationStyle = R.style.showAsDropDown
        //防止被虚拟导航栏阻挡
        this.softInputMode = WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE

        initView(parent)
    }

    open fun getSelfTitleView(): Int {
        return 0
    }

    abstract fun getSelfContentView(): Int

    open fun getSelfButtonView(): Int {
        return 0
    }

    abstract fun initView(parent: View)

    /**
     * description 数据渲染，放在子类构造方法的super()之后执行，避免构造方法传参无法获取的问题
     */
    abstract fun initData()

    open fun setTitle(view: View?) {
        mTitle?.removeAllViews()
        if (view != null) {
            mTitle?.addView(view)
        }
    }

    open fun setContent(view: View?) {
        mContent?.removeAllViews()
        if (view != null) {
            mContent?.addView(view)
        }
    }

    open fun setButton(view: View?) {
        mButton?.removeAllViews()
        if (view != null) {
            mButton?.addView(view)
        }
    }

    override fun getContentView(): View? {
        return super.getContentView()
    }

    override fun setContentView(view: View?) {
        super.setContentView(view)
    }

    override fun setAnimationStyle(animationStyle: Int) {
        super.setAnimationStyle(animationStyle)
    }

    override fun setBackgroundDrawable(background: Drawable?) {
        super.setBackgroundDrawable(background)
    }

    override fun showAsDropDown(view: View?) {
        super.showAsDropDown(view)
    }

    /**
     * @description 向下弹出
     */
    open fun setDropDown() {
        //设置SelectPicPopupWindow弹出窗体动画效果
        this.animationStyle = R.style.showAsDropDown
    }

    /**
     * @description 向上弹出
     */
    open fun setDropUp() {
        this.animationStyle = R.style.showAsDropUp
    }

    /**
     * @description  向左弹出
     */
    open fun setDropLeft() {
        this.animationStyle = R.style.showAsDropLeft
    }

    /**
     * @description向右弹出
     */
    open fun setDropRight() {
        this.animationStyle = R.style.showAsDropRight
    }

    override fun showAtLocation(parent: View?, gravity: Int, x: Int, y: Int) {
        super.showAtLocation(parent, gravity, x, y)
    }

    /**
     * 窗口变暗
     */
    open fun showBlackWindowBackground(activity: Activity?) {
        activity?.let {
            val lp = activity.window?.attributes
            lp!!.alpha = 0.4f
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
            activity.window.attributes = lp
        }
    }

    open fun dismissBlackWindowBackground(activity: Activity?) {
        activity?.let {
            val lp = activity.window.attributes
            lp.alpha = 1f
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
            activity.window.attributes = lp
        }
    }

    /**
     * 取消监听器
     */
    interface OnCancelClickListener {
        fun onCancel(view: View?)
    }

    /**
     * 确定监听器
     */
    interface OnConfirmClickListener {
        fun onConfirm(view: View?)
    }

    init {
        initView(context)
    }
}